### PR TITLE
[DCP][DSD] Fix to remove non_persistent buffer in distributed state dict

### LINKED ISSUE #122792: [DCP] `set_model_state_dict` errors on compiled module with non-persistent buffer
```
"""
torchrun --standalone --nproc_per_node=2 repro_dcp_compile.py
"""
import os
import torch
import torch.nn as nn
import torch.distributed as dist
from torch.distributed.checkpoint.state_dict import get_model_state_dict, set_model_state_dict

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin1 = nn.Linear(4, 4)
        self.lin2 = nn.Linear(4, 4)
        self.register_buffer("buf", torch.randn((4,)), persistent=False)
        self.weight = nn.Parameter(torch.randn((4, 4)))


if __name__ == "__main__":
    dist.init_process_group(backend="nccl")
    gpu_id = int(os.environ["LOCAL_RANK"])
    device = f"cuda:{gpu_id}"
    torch.cuda.set_device(device)

    model = Model()
    model = torch.compile(model)

    sharded_sd = get_model_state_dict(model)
    set_model_state_dict(model, sharded_sd)
```

```
[rank0]: Traceback (most recent call last):
[rank0]:   File "/data/users/andgu/pytorch/repro_dcp_compile.py", line 36, in <module>
[rank0]:     set_model_state_dict(model, sharded_sd)
[rank0]:   File "/data/users/andgu/pytorch/torch/distributed/checkpoint/state_dict.py", line 853, in set_model_state_dict
[rank0]:     return _load_model_state_dict(model, model_state_dict, info)
[rank0]:   File "/data/users/andgu/pytorch/torch/distributed/checkpoint/state_dict.py", line 416, in _load_model_state_dict
[rank0]:     state_dict[fqn_with_prefix] = state_dict.pop(fqn)
[rank0]: KeyError: 'buf'
```

`set_model_state_dict` calls into `_load_model_state_dict`, which iterates over `named_buffers()`. For a compiled module, `fqns` and `fqns_with_prefix` always mismatch, so `_load_model_state_dict` will try to reassign from the FQN without prefix to the one with prefix. However, this does not account for non-persistent buffers not existing in the state dict.

One solution could be just to continue `if fqn not in state_dict`.

cc @LucasLLC

### PR DESCRIPTION
**Pull request by wz337**
Fixes #122792

`state_dict` includes only persistent buffers, while `named_buffers()` would include non_persistent buffers.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @chauhang @LucasLLC

### CODE REVIEW

**Review by awgu — APPROVED** [review]
Thanks for the super fast PR!

File: torch/distributed/checkpoint/state_dict.py [hunk 1]
```diff
@@ -121,9 +121,9 @@ class StateDictOptions:
 
 @dataclass
 class _StateDictInfo(StateDictOptions):
-    fqn_param_mapping: Dict[
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/checkpoint/state_dict.py:124]: Ye. Linter was complaining about it.

**Review by fegin — APPROVED** [review]
The reason I did not use `model.state_dict()` is that if the model is parallelized by FlatParameter FSDP, this will incur an extra communication. Not sure if this will cause a performance issue.

**Comment by wz337** [comment]
> The reason I did not use `model.state_dict()` is that if the model is parallelized by FlatParameter FSDP, this will incur an extra communication. Not sure if this will cause a performance issue.

Found an internal field of nn.module that let me check whether a buffer is non_persistent_buffers so we can avoid the state_dict call for FlatParameter FSDP. So instead of using `state_dict`, falling back to use `named_buffers()` but skip it if it is non persistent.

**Pull request closed without merging**